src/utils/reporting_workflow.js

```diff
@@ -0,0 +1,154 @@
+ */
+const IG_REGISTRY = {
+  [BASE_REPORTING_BUNDLE_PROFILE]: {} // base action definitions, see MEDMORPH-35
+};
+
+function initializeContext(planDefinition) {
+  // TODO -- see MEDMORPH-36
+  return { id: uuidv4(), planDefinition };
+}
+
+function getFunction(ig, code) {
+  // use a function here in case this needs to be more complex
+  return IG_REGISTRY[ig][code];
+}
+
+/**
+ * Find the target profile to be used when generating a bundle
+ * in support of the given PlanDefinition.
+ * The profile should be on the 'output' field of the 'create-report' action.
+ */
+function findProfile(planDefinition) {
+  // figure out the profile of the target bundle
+
+  // TODO search arrays rather than hardcode 0,0
+  const createReportAction = planDefinition.action?.find(
+    a => a.code[0].coding[0].code === 'create-report'
+  );
+
+  if (!createReportAction) {
+    throw new Error(`PlanDefinition ${planDefinition.id} has no action with code 'create-report'`);
+  }
+
+  const bundleOutput = createReportAction.output?.find(o => o.type === 'Bundle');
+
+  if (!bundleOutput?.profile?.[0]) {
+    throw new Error(
+      `PlanDefinition ${planDefinition.id} does not specify a profile for report bundle`
+    );
+    // option B is just use the default profile
+  }
+
+  return bundleOutput.profile[0];
+}
+
+function determineActionSequence(planDefinition) {
+  // TODO:
+  // note: 6.1.8
+  // The Backend Service App SHALL implement the ability
+  // to process action dependencies specified via
+  // the PlanDefinition.action.relatedAction elements.
+  // Typical implementations will build a graph or a tree structure
+  // that can be used to identify dependencies between actions.
+
+  // simple version to just get the integers 0 -> n
+  // https://stackoverflow.com/a/10050831
+  return [...Array(planDefinition.action.length).keys()];
+}
+
+function executeWorkflow(context) {
+  if (!context.currentActionSequenceStep) {
+    context.currentActionSequenceStep = 0;
+  }
+
+  const planDef = context.planDefinition;
+
+  if (!context.actionSequence) {
+    context.actionSequence = determineActionSequence(planDef);
+  }
+
+  db.upsert(COLLECTION, context, c => c.id === context.id);
+
+  const ig = findProfile(planDef);
+
+  while (context.currentActionSequenceStep < context.actionSequence.length) {
+    const currentActionId = context.actionSequence[context.currentActionSequenceStep];
+    const action = planDef.action[currentActionId];
+
+    // TODO - check arrays for
+    // system === "http://hl7.org/fhir/us/medmorph/CodeSystem/us-ph-plandefinition-actions"
+    const actionCode = action.code[0].coding[0].code;
+
+    const execute = getFunction(ig, actionCode);
+
+    // TODO: maintain context at each step for debugging?
+    // context = deepCopy(context);
+
+    // check the DB to get cancelToken
+    // TODO: is there a better way to do this?
+    const cancelToken = db.select(COLLECTION, c => c.id === context.id).cancelToken;
+    if (cancelToken) return;
```

Comment:
Does this need to use the db? When would the stored context be different from the context passed to this function?

Also, nothing in the for loop is required for this so it can be moved up to the top of this function.

Reply:
The implementation of the cancelToken is admittedly half-baked, at best, and very possibly unnecessary. The idea is that it's something that could be set from elsewhere, in order to terminate this running process. For example, imagine processing one notification for a patient, then a second notification for the same patient comes in right away. Maybe we have the second job check to see if there is a duplicate process running and have it kill the first one, so only one report gets sent with the latest data. So one way to do that is with a cancellation token, which the process checks before every step. So the way for another process to set that is via the database. 

The idea comes from C#/.NET: https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads